### DISCUSSION TITLE
Programmatically detect transformed column names which can be passed to ForecastX in transformer+forecaster pipeline

### DISCUSSION THREAD

**Discussion by yarnabrina** [alan-turing-institute/sktime#4830] [Q&A]
Let's confider a forecasting problem with one or many endogenous variables, and at least two exogenous variables. At least one of these exogenous variables are unknown for the forecasting horizon, and needs to be forecasted first before predicting for the endogenous variable. And, at least one of the other exogenous variable is known by some other means for the prediction duration.

This works on its own, but now I'm trying to add few transformations before going into forecasting stage, and `FeatureUnion` seems to change the column names. But since `ForecastX` needs the specification of columns that need to be predicted in advance (as I have at least one exogenous feature for which I shall provide the values myself during `predict`).

This is a code to demonstrate my problem:

```python
from sktime.datasets import load_longley
y, X = load_longley()

future_known_exogenous = ["GNPDEFL", "GNP"]
future_unknown_exogenous = ["UNEMP", "ARMED", "POP"]

from sktime.forecasting.base import ForecastingHorizon
fh = ForecastingHorizon([1, 2, 3, 4])

from sktime.forecasting.model_selection import temporal_train_test_split
y_train, y_test, X_train, X_test = temporal_train_test_split(y, X=X, fh=fh)

from sklearn.preprocessing import RobustScaler
sklearn_scaler = RobustScaler()

from sklearn.preprocessing import QuantileTransformer
sklearn_transformer = QuantileTransformer()

from sktime.transformations.series.adapt import TabularToSeriesAdaptor
sktime_scaler = TabularToSeriesAdaptor(sklearn_scaler)
sktime_transformer = TabularToSeriesAdaptor(sklearn_transformer)

from sktime.transformations.series.subset import ColumnSelect
future_known_selector = ColumnSelect(columns=future_known_exogenous)
future_unknown_selector = ColumnSelect(columns=future_unknown_exogenous)

from sktime.transformations.compose import TransformerPipeline
future_known_transformer = TransformerPipeline([future_known_selector, sktime_scaler])
future_unknown_transformer = TransformerPipeline([future_unknown_selector, sktime_transformer])

from sktime.transformations.compose import FeatureUnion
transformer = FeatureUnion([future_known_transformer, future_unknown_transformer])

from sktime.transformations.series.detrend import Detrender
de_trender = Detrender()

from sktime.transformations.series.detrend import Deseasonalizer
de_seasonaliser = Deseasonalizer()  # will prefer to pass significant sp from SeasonalityACF (how?)

from sktime.forecasting.arima import ARIMA
algorithm = ARIMA()

from sktime.forecasting.compose import TransformedTargetForecaster
model = TransformedTargetForecaster([de_trender, de_seasonaliser, algorithm])

from sktime.forecasting.compose import ForecastX
forecaster = ForecastX(model.clone(), model.clone(), columns=future_unknown_exogenous)

from sktime.forecasting.compose import ForecastingPipeline
pipeline = ForecastingPipeline([transformer, forecaster])

pipeline.fit(y_train, X=X_train, fh=fh)  # fails with KeyError
# KeyError: "None of [Index(['UNEMP', 'ARMED', 'POP'], dtype='object')] are in the [columns]"

pipeline.predict(fh=fh, X=X_test)  # objective
```

In this example, I wish to run a subset of columns through scaling and another subset through transformation. These subsets match with the subsets of future know/unknown subsets to simplify illustration, and that need not be the case. Also, all these transformations are series-to-series type, and in a general case can very well be series-to-dataframe type, for example `FourierFeatures`. I also pass the target variables through standard removal of non-stationarity process before passing to the forecasting algorithm.

The transformer and the forecaster works on their own, but as I said `FeatureUnion` seems to add a prefix to the column names which do not match the names I have access to in advance while defining the forecaster, and it's failing after complaining that columns are missing.

So, my main question is how to access the output column types from the transformer pipeline and pass to the forecaster pipeline automatically so that I do not have to worry about doing separately and add aditional inverse transformation calls.

A side question that I came up with while creating the example is if it'd be possible to pass `sp` from `SeasonalityACF` and pass to `Deseasonalizer` or not. It's very much related to the `auto_detect_sp` discussion, but not exactly same as here I wish to add it through pipeline and estimator modification does not come into picture.

**Comment by yarnabrina** [comment]
FYI @fkiraly @davidgilbertson

**Comment by davidgilbertson** [comment]
I had a similar problem. In my case, I could safely remove all the added column prefixes (this didn't result in duplicate column names).

I have a function like this:

```py
def clean_cols(df):
    df.columns = df.columns.str.replace(r"^ColumnSelect_\d+__", "", regex=True)
    df.columns = df.columns.str.replace(r"^ColumnSelect__", "", regex=True)
    df.columns = df.columns.str.replace(r"^TransformerPipeline__", "", regex=True)
    df.columns = df.columns.str.replace(r"^TransformerPipeline_\d+__", "", regex=True)
    df.columns = df.columns.str.replace(r"_lag_0$", "", regex=True)

    if not df.columns.is_unique:
        print("Careful, you've got duplicate columns:", df.columns)

    return df
```

Then in the `TransformerPipeline`, after a transformer step that changes column names, I add the step:

```py
FunctionTransformer(clean_cols)
```

It's quite hacky/brittle but worked for my use case.

**Reply by yarnabrina** [reply]
Thanks, but yes, it's very hacky. For my example above, it's okay I guess (after I modify the regex), but not in general.

I think it'll only work if each column gets transformed at most once (counting sequence as one) and finally becomes a single series. I'm not sure if this restriction is intended, may be @fkiraly can confirm.

**Reply by fkiraly** [reply]
See my general answer here: https://github.com/sktime/sktime/discussions/4830#discussioncomment-6376085

More specifically, about the column name handling, I was thinking of introducing to every estimator that can potentially generate nested column names and argument `output_cols` or similar, with options:

* `multiindex` - `pd.MultiIndex` with column levels added, e.g., `("trafoname", "varname")`
* `flat_multiindex` - the above flatted, e.g., `"trafoname__varname"`
* `original` - the original variable name, e.g., `varname`. This is made unique by appending numbers if there are clashes, and raises a warning
* `auto` - same as from `original` if unique, but replaced by `flat_multiindex` if non-unique.

**Reply by fkiraly** [reply]
default to be `auto` in end state and `flat_multiindex` (current default) during deprecation period.

**Comment by yarnabrina** [comment]
Hi @davidgilbertson, I've a follow-up question, which is related to the flow but not to the issue in my original question.

I'm observing that I can not call the `predict` method of `pipeline` without passing a `X` argument. But in cases where transformation is required only for training, for example when all `X` are unknown in future and `ForecastX` should handle it itself, it fails complaining `X` can not be `None`.

I'm not sure if it's a bug, as `transform` method is called for transformers without anything to transform, but I want to how it's handled usually. Have you ever faced this?

**Reply by fkiraly** [reply]
Yes, also when building pipelines and using `ForecastX` - that's why I drafted this a while ago:
https://github.com/sktime/sktime/pull/4112
review appreciated!

**Reply by yarnabrina** [reply]
@fkiraly Given that I'm currently being bothered by this in my work, I'm super biased to have this support. However, given my <2 days of experience with `sktime` transformers, it's better to refrain from any comment on design questions.

(Just to be clear, I also do not have any comments as it is currently.)

---

Do you have any suggestions on the column name detection, the main issue above? As of now I am doing what @davidgilbertson suggested, but this workaround is significantly decreasing the set of possible transformations while using with `ForecastX`.

**Reply by fkiraly** [reply]
> I'm super biased to have this support.

Review appreciated :-)

> Do you have any suggestions on the column name detection, the main issue above

Yes, but will need some time to write this up properly. Will do so after work.

**Comment by fkiraly** [comment]
@yarnabrina, my response and suggested actions to this would be as follows:

* I think the "column renaming" problem is the same that @davidgilbertson had
* my primary reaction was this PR that updates `ColumnTransformer` with an option to keep the column names: https://github.com/sktime/sktime/pull/4789
* as a secondary measure, I would also suggest to introduce the same `output_names` parameter to `FeatureUnion`, to replace `flatten_multiindex` with at least a third option to retain the original column names.

**Reply by fkiraly** [reply]
the quickest way to get somewhere imo with a general solution would be to finalize #4789 (?), reviews appreciated

**Reply by fkiraly** [reply]
& 0.20.2 released on the week-end?

**Comment by davidgilbertson** [comment]
@yarnabrina another option (short-term workaround) that's worked for me: I have a pipeline where I wanted to drop some columns (and avoid the column-renaming issues), I figured this would be something I do often enough, so I just made my own transformer, it's not so hard:
```py
class DropColumns(FunctionTransformer):
    def __init__(self, cols: Sequence[str] | str):
        super().__init__()
        if isinstance(cols, str):
            cols = [cols]
        self.cols = cols

    def _fit(self, X, y=None):
        return self

    def _transform(self, X, y=None):
        return X.drop(columns=self.cols)
```

Basically, if you know how to do what you want to do in Pandas, you can just put it in the `_transform` method of a transformer. So, for example you could make your own `FeatureUnion` or even put all your logic in your own transformer that doesn't change column names.

**Reply by yarnabrina** [reply]
@davidgilbertson I don't think I can leverage this for my use case. If I knew what (modified) columns is available, then I could have used in ForecastX. Problem is I don't seem to get that information programmatically.

In a particular problem, sure I can run once, let it fail but note the names, and use correctly in second time. But my current use case is we are creating a solution on top of sktime where end users (non technical folks) will pass any data and I have to accommodate those. So, I don't think I'll be able to use this approach.

**Reply by fkiraly** [reply]
minor comments on the `DropColumns`:

* if it is a plain `pandas` function, you can use `PandasTransformAdaptor`
* is this not the same as `ColumnSubset`, just with the opposite set of columns? Maybe worth adding to `sktime`?
* you should check your transformer with `check_estimator` on a good selection of parameter settings, in this case you should test at least `cols` being iterable and being `str`.
* In the `str` case, it will fail, because you are violating the `scikit-learn` contract that `self.param` should be the arg of `__init__` always.

**Reply by davidgilbertson** [reply]
Thanks @fkiraly I hadn't seen `ColumnSubset` - did you know it doesn't show up in the docs?

**Reply by fkiraly** [reply]
@davidgilbertson, apologies, it's called `ColumnSelect` (not `ColumnSubset`) and should be in one of the sections in there.
I just checked and I think it's visible in the docs.

Do you think it should be better in another place?

**Comment by yarnabrina** [comment]
Hi @fkiraly, I was taking a look back at this again, and I am a bit confused how to do this now that #4112 and #4789 are both merged. To use `ForecastX` on only few columns, I somehow need to find which transformed columns come from those, and have to allow for the possibility of series->dataframe transformations (e.g. one-hot).

(FYI: @davidgilbertson / @benHeid since I think you are also using these.)

**Reply by fkiraly** [reply]
yes, this is still not 100% straightforward - I wanted to write a STEP to add a `get_feature_names_out` method like in `sklearn`
